### ISSUE TITLE
GUI backend not found

### ISSUE THREAD

**Issue by bkong** [stevengj/PyPlot.jl#110]
I have matplotlib installed and it works in Python. A simple test like:
```
import matplotlib.pyplot as plt
plt.plot([1,2,3,4])
plt.show()
```
works just fine.

However, PyPlot is not able to locate a working GUI backend for matplotlib. Specificially I get
```
WARNING: No working GUI backend found for matplotlib.
```
I'm using matplotlib-1.3.0 on Gentoo Linux with the GTK backend installed. Any ideas on what the problem is and how to fix this?

**Comment by stevengj** [comment]
What happens if you specifically tell it to use GTK?  i.e.
```
using PyCall
pygui(:gtk)
using PyPlot
```

**Comment by stevengj** [comment]
Also, what does
```
using PyCall
get(PyDict(pyimport("matplotlib")["rcParams"]), "backend", "none")
```
return?

(A common problem to watch out for is if you have multiple versions of Python installed and PyCall is not using the one you want.  You can force it to use a particular Python by setting the `PYTHON` environment variable, before running Julia, e.g. run `PYTHON=python3 julia`)

**Comment by bkong** [comment]
Running pygui(:gtk) results in
```
ERROR: Python GUI toolkit for gtk is not installed.
```
Running the second chunk returns  "GTK3Cairo". This is true regardless of whether I run using PYTHON=python3 or not. However, if I run those commands using PYTHON=python2, I get "GTK3Agg". Does the problem lie in that it is reporting GTK3 rather than just GTK?

**Comment by stevengj** [comment]
That error is bad news: it means that `pyimport("gtk")` is failing.   Run
```julia
using PyCall
pyimport("gtk")
```
to see what exception is being thrown by Python.

**Comment by bkong** [comment]
With ```PYTHON=python2``` I get
```
PyObject <module 'gtk' from '/usr/lib64/python2.7/site-packages/gtk-2.0/gtk/__init__.pyc'>
```
With ```PYTHON=python3``` I get
```
ERROR: PyError (:PyImport_ImportModule) <class 'ImportError'>
ImportError("No module named 'gtk'",)
```

**Comment by bkong** [comment]
Just wanted to add what happens when I call ```using PyPlot``` with ```PYTHON=python2```:
```
/usr/lib64/python2.7/site-packages/matplotlib/__init__.py:1141: UserWarning:  This call to matplotlib.use() has no effect
because the the backend has already been chosen;
matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.

  warnings.warn(_use_error_msg)
WARNING: No working GUI backend found for matplotlib.
Warning: error initializing module PyPlot:
PyCall.PyError(msg=":PyImport_ImportModule", T=PyCall.PyObject(o=0x0000000068648320), val=PyCall.PyObject(o=0x0000000054ba6830), traceback=PyCall.PyObject(o=0x0000000054b9af38))
```

**Comment by bkong** [comment]
Seems like using PyPlot with GTK+-3.x won't be possible since PyGTK only supports GTK+-2.x, hence there was no module named ```'gtk'``` when I was testing with ```PYTHON=python3```. Now with ```PYTHON=python2``` the issue was an encoding issue that was fixed upstream a while ago, but since Gentoo Linux uses an ancient version (1.3.0) as their stable version I didn't have the fix.

tl;dr Everything is working now with python2 and matplotlib-1.4.2. =)

**Comment by stevengj** [comment]
GTK3 support should be fixed by bb24ff8b8f716845094a0622a46c57960b392a45 ... @bkong, please try it and let me know.

**Comment by bkong** [comment]
Seems to be fixed. I was getting the following warning using the latest PyPlot,
```
/usr/lib64/python3.3/site-packages/matplotlib/backends/backend_gtk3agg.py:18: UserWarning: The Gtk3Agg backend is known to not work on Python 3.x with pycairo. Try installing cairocffi.
  "The Gtk3Agg backend is known to not work on Python 3.x with pycairo. "
```
but everything is now working after installing cairocffi as the warning suggested.

**Comment by davidzentlermunro** [comment]
I'm having similar issues and following instructions on this and similar posts does not seem to have helped. Here is the error code I get after I try `using PyPlot`:

`WARNING: No working GUI backend found for matplotlib.
ERROR: InitError: PyError (:PyImport_ImportModule) <class 'ImportError'>
ImportError("/lib64/libz.so.1: version ZLIB_1.2.3.4 not found (required by /data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/../../../libpng16.so.16)",)
  File "/data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/pyplot.py", line 29, in <module>
    import matplotlib.colorbar
  File "/data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/colorbar.py", line 34, in <module>
    import matplotlib.collections as collections
  File "/data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/collections.py", line 27, in <module>
    import matplotlib.backend_bases as backend_bases
  File "/data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/backend_bases.py", line 62, in <module>
    import matplotlib.textpath as textpath
  File "/data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/textpath.py", line 18, in <module>
    from matplotlib.mathtext import MathTextParser
  File "/data/uctpdtz/.julia/anaconda3/lib/python3.5/site-packages/matplotlib/mathtext.py", line 60, in <module>
    import matplotlib._png as _png

 [inlined code] from /data/uctpdtz/.julia/v0.4/PyCall/src/exception.jl:81
 in pyimport at /data/uctpdtz/.julia/v0.4/PyCall/src/PyCall.jl:79
 in __init__ at /data/uctpdtz/.julia/v0.4/PyPlot/src/PyPlot.jl:224
 in _require_from_serialized at loading.jl:84
 in _require_from_serialized at ./loading.jl:109
 in require at ./loading.jl:219
during initialization of module PyPlot`

**Comment by stevengj** [comment]
@davidzentlermunro, that sounds like an unrelated issue. Sounds like you have a missing libz library (or a missing library path) in your matplotlib install.   My usual recommendation is to [use the automated Conda-based installation](https://github.com/stevengj/PyPlot.jl#automated-matplotlib-installation).

**Comment by davidzentlermunro** [comment]
@stevengj  I tried doing this by running `ENV["PYTHON"]=""` and then building PyCall again (as per instructions) - it worked while I was running the interactive session but when I exited and reloaded Julia `using PyPlot` failed again

**Comment by stevengj** [comment]
That's really odd.   The `PYTHON` environment variable shouldn't affect anything outside of `Pkg.build`, but I suppose you could try to set it before `using PyPlot` to see if that changes anything?

**Comment by davidzentlermunro** [comment]
I realised I didn't give the full error message previously. Here it is:

`/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/__init__.py:1350: UserWarning:  This call to matplotlib.use() has no effect
because the backend has already been chosen;
matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.
  warnings.warn(_use_error_msg)
WARNING: No working GUI backend found for matplotlib.
ERROR: InitError: PyError (:PyImport_ImportModule) <type 'exceptions.ImportError'>
ImportError("/lib64/libz.so.1: version `ZLIB_1.2.3.4` not found (required by /data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/../../../libpng16.so.16)",)
  File "/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/pyplot.py", line 29, in <module>
    import matplotlib.colorbar
  File "/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/colorbar.py", line 34, in <module>
    import matplotlib.collections as collections
  File "/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/collections.py", line 27, in <module>
    import matplotlib.backend_bases as backend_bases
  File "/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 62, in <module>
    import matplotlib.textpath as textpath
  File "/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/textpath.py", line 18, in <module>
    from matplotlib.mathtext import MathTextParser
  File "/data/uctpdtz/.julia/v0.4/Conda/deps/usr/lib/python2.7/site-packages/matplotlib/mathtext.py", line 60, in <module>
    import matplotlib._png as _png
 [inlined code] from /data/uctpdtz/.julia/v0.4/PyCall/src/exception.jl:81
 in pyimport at /data/uctpdtz/.julia/v0.4/PyCall/src/PyCall.jl:79
 in __init__ at /data/uctpdtz/.julia/v0.4/PyPlot/src/PyPlot.jl:224
 in _require_from_serialized at loading.jl:84
 in _require_from_serialized at ./loading.jl:109
 in require at ./loading.jl:219
during initialization of module PyPlot`

**Comment by stevengj** [comment]
@davidzentlermunro, you are typing `using PyPlot` and nothing else when you start your Julia session?

**Comment by davidzentlermunro** [comment]
Correct. One more weird thing: if I type `using PyPlot` again (after that error message) it works fine (though I get the same error message if I quit and start again).
